### PR TITLE
Store actual record size in data entry and refactor pmem allocator

### PR DESCRIPTION
**Pull request by JiayuZzz**
### What problem does this PR solve?

Problem Summary:

Record size stored in data entry is in block unit in current code, which can be confused and is coupling with pmem allocator

### What is changed and how it works?

What's Changed:

Store actual size of record in data entry, and refactored pmem allocator to adjust to it

### Check List 

Tests 

- Unit test
- Stress test
- Benchmark test

Side effects

no

### CODE REVIEW

File: engine/kv_engine.cpp [hunk 1]
```diff
@@ -792,8 +792,10 @@ Status KVEngine::HashGetImpl(const pmem::obj::string_view &key,
     }
 
     // Copy PMem data record to dram buffer
-    // TODO jiayu: record_size maybe invalid here?
-    auto record_size = data_entry.header.record_size * configs_.pmem_block_size;
+    auto record_size = data_entry.header.record_size;
+    if (record_size > configs_.pmem_segment_blocks * configs_.pmem_block_size) {
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:796]: Is this for a merged reused space?
please add comments.

File: engine/kv_engine.cpp [hunk 2]
```diff
@@ -846,8 +848,8 @@ Status KVEngine::SDeleteImpl(Skiplist *skiplist,
     auto hint = hash_table_->GetHint(collection_key);
     std::lock_guard<SpinMutex> lg(*hint.spin);
     Status s =
-        hash_table_->SearchForWrite(hint, collection_key, SortedDataRecord,
-                                    &entry_ptr, &hash_entry, &data_entry);
+        hash_table_->SearchForRead(hint, collection_key, SortedDataRecord,
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:851]: SearchForInsert
Search // for read, delete and update

Comment (JiayuZzz) [reply engine/kv_engine.cpp:851]: rename in other patch

File: engine/kv_engine.cpp (hunk 1 above)
Comment (JiayuZzz) [reply engine/kv_engine.cpp:796]: done

**Review by ZiyanShi — APPROVED** [review]

**Pull request merged**